src/esp32/hcsr04.rs

```diff
@@ -0,0 +1,289 @@
+                anyhow::anyhow!(
+                    "HCSR04Sensor: failed to get `echo_interrupt_pin` from configuration: {:?}",
+                    e
+                )
+            })?;
+
+        let timeout = cfg.get_attribute::<u32>("timeout_ms").map_or_else(
+            |e| match e {
+                AttributeError::KeyNotFound(_) => Ok(None),
+                _ => Err(anyhow::anyhow!(
+                    "HCSR04Sensor: error handling `timeout_ms` value: {:?}",
+                    e
+                )),
+            },
+            |v| Ok(Some(Duration::from_millis(v as u64))),
+        )?;
+
+        Ok(Arc::new(Mutex::new(HCSR04Sensor::new(
+            board,
+            trigger_pin,
+            echo_interrupt_pin,
+            timeout,
+        )?)))
+    }
+
+    fn new(
+        board: BoardType,
+        trigger_pin: i32,
+        echo_interrupt_pin: i32,
+        timeout: Option<Duration>,
+    ) -> anyhow::Result<HCSR04Sensor> {
+        // TODO: Move this into micro-rdk and unify with pin.rs
+        init_isr_alloc_flags(esp_idf_hal::interrupt::InterruptType::Iram.into());
+        enable_isr_service()?;
+
+        let notification = Notification::new();
+        let notifier = notification.notifier();
+
+        let mut sensor = Self {
+            _board: board,
+            trigger_pin: PinDriver::output(unsafe { AnyIOPin::new(trigger_pin) })?,
+            echo_interrupt_pin: PinDriver::input(unsafe { AnyIOPin::new(echo_interrupt_pin) })?,
+            timeout: timeout.unwrap_or(Duration::from_millis(50)),
+            interrupt_notification: notification,
+            isr_shared_state: Arc::new(IsrSharedState {
+                timestamp: 0.into(),
+                notifier,
+            }),
+        };
+
+        sensor.echo_interrupt_pin.set_pull(Pull::Down)?;
+
+        // Start the trigger pin high: the pulse is sent on the
+        // falling edge, so we can just go low immediately in
+        // `get_readings` to send it. As long as we don't get another
+        // `get_readings` request within 10us of the prior one
+        // completing, the pin will be high long enough to trigger the
+        // pulse.
+        sensor.trigger_pin.set_high()?;
+
+        sensor
+            .echo_interrupt_pin
+            .set_interrupt_type(InterruptType::AnyEdge)?;
+
+        unsafe {
+            esp!(gpio_isr_handler_add(
+                echo_interrupt_pin,
+                Some(Self::subscription_interrupt),
+                Arc::as_ptr(&sensor.isr_shared_state) as *mut _,
+            ))?;
+        }
+
+        Ok(sensor)
+    }
+
+    #[inline(always)]
+    #[link_section = ".iram1.intr_srv"]
+    unsafe extern "C" fn subscription_interrupt(arg: *mut core::ffi::c_void) {
+        let arg: &mut IsrSharedState = &mut *(arg as *mut _);
+        let when = esp_idf_sys::esp_timer_get_time();
+        match arg
+            .timestamp
+            .compare_exchange(0, when, Ordering::AcqRel, Ordering::Acquire)
```

Comment:
I am not sure you need to atomically store the timestamp result with the way the code is written since it can only be updated from the ISR and read by the thread after it was notified by the ISR.

Reply:
I had a different question here: why doesn't the atomic nature of the timestamp cover the case that the lockout state is trying to guard against? (I know you were considering removing lockout anyway, so if you do please disregard)

Reply:
also @npmenard , doesn't `get_readings` try to access it outside of an ISR context (L. 251)? Or will that still not interfere?

Reply:
@npmenard - My reasoning in favor of an atomic here is that the timestamp gets written by the first interrupt (rising edge of echo pin) and the value written then needs to be read by the second interrupt (falling edge of echo pin) in order to compute the delta between them. As far as I understand there needs to be a happens-before edge between the write and the read, and without the atomic there would not be. It also isn't clear to me that the notification enforces any global ordering such that a non-atomic write in the ISR would be seen by the read in `get_readings`, ditto the write of zero to "arm" the interrupt being seen in the ISR, as noted by @gvaradarajan.

@gvaradarajan - I'm about 90% sure that the lockout isn't needed, I just didn't want to hold up the review  until I've re-tested, since if it isn't actually needed its just a simplification. What I was observing was what looked like spurious edges when the sensor timed out. But things also weren't fully debugged at that point and it is entirely possible that those "spurious edges" were in fact my code misbehaving. Some testing should quickly make it clear.

Reply:
I see, I was reasoning that since interrupts cannot be pre-empted then all writes would flush when exiting the interrupt context (since all registers would be swapped) and since measuring the with ultrasonic has strongly ordered steps then  you couldn't get where you read a variable  in one thread when an interrupt comes in. However I couldn't find anything backing the first assumption in the technical reference for Xtensa and Esp32 plus I missed that the variable is i64 so no atomic read/write by default so I think it is better to leave the implementation  as is